### PR TITLE
add core.on_place, core.on_punch, can_place, can_punch, “fix code style“, use assert instead of error, remove unused, etc…

### PR DESCRIPTION
**Pull request by HybridDog**
See also http://dev.minetest.net/User:Hybrid_Dog#minetest.on_place

### CODE REVIEW

**Comment by PilzAdam** [comment]
Please don't mix style fixes with new functionality.

Also the doc is a bit misleading: `if it doesn't return true` what is "it?" The function being called or the function that is passed as a parameter?

File: builtin/game/misc.lua [hunk 1]
```diff
@@ -99,33 +101,39 @@ core.register_on_leaveplayer(function(player)
...
 end)
 
+-- works faster than get_connected_players but doesn't test if it's connected
+function core.get_player_list()
+	return player_list
+end
+
 function core.get_connected_players()
-	local temp_table = {}
+	local temp_table,n = {},1
```
Comment (rubenwardy) [inline builtin/game/misc.lua]: I don't find this particularly easy to read.

Comment (HybridDog) [inline builtin/game/misc.lua]: sorry

**Comment by HybridDog** [comment]
>Please don't mix style fixes with new functionality.

est wrote that l should do it
https://github.com/minetest/minetest/pull/2592#discussion_r27773572
>Also the doc is a bit misleading: if it doesn't return true what is "it?" The function being called or the function that is passed as a parameter?

fixed

**Comment by est31** [comment]
I've only said if you actually change something in a code line, then you should fix its style.
I have not said that you should randomly fix the style in the whole file.

**Comment by est31** [comment]
1. `can_place`, and `can_punch` should get the the candidate position passed, see #3133. This is important as people often think pointed_thing.above always denotes the candidate position. This is wrong however, [under special conditions](https://github.com/minetest/minetest/blob/2b04ab874d75711bc021a0cd8dc7fca68f4e6929/builtin/game/item.lua#L224-L228). Therefore I think its easiest to just pass the candidate position. pointed_thing can still be passed as well.

2. The naming of `on_place` and `on_punch` is bad. better would be `override_on_place` or so.

3. I don't really see the reason why `on_place` or `on_punch` would be useful.

**Comment by HybridDog** [comment]
>I've only said if you actually change something in a code line, then you should fix its style.
I have not said that you should randomly fix the style in the whole file.

here l'm changing code at many places in the file, e.g. the error→assert replacements, so l think the code style fix belongs to this
> This is wrong however, under special conditions.

thanks for giving this link, l'll try to use the info the code gives there
>Therefore I think its easiest to just pass the candidate position, so that you don't have to replicate that code yourself. pointed_thing can still be passed as well.

ok, pos needs to be a copied vector, doesn't it? l guess people maybe would do e.g. pos.y = pos.y-1 and wonder why pointed thing changed.
>The naming of on_place and on_punch is bad. better would be override_on_place or so.

ok, l guess did it that the code looks similar to the e.g. on_destruct and after_destruct
>I don't really see the reason why core.on_place or core.on_punch would be useful.

you can use them to
* set an on_punch of a node without disabling support for mods on this node, e.g. superpick, worldedit
* disallowing placing e.g. wood into lava, torches into water, etc., which is better than overriding the node's can_place because if another mod also wants to override it…

**Comment by t4im** [comment]
> here l'm changing code at many places in the file, e.g. the error→assert replacements,
> so l think the code style fix belongs to this

One significant change — one commit.

If they wanted to apply your style fixes but not the error->assert, they would only be able to do that, if you committed them separately. Fixing style in a line or statement you touch anyway can usually just be considered part of the "change done right". But randomly fixing all sorts of stuff belongs in a style-fix or whitespace-fix commit. 

Also if someone would have to revert the error->assert another time, they might want to do so separately from style fixes as well.

Which brings me to the asserts.
Let me quote a quick explanation of their purpose in relation to errors:

> Assertions are distinct from routine error-handling. Assertions document logically impossible situations and discover programming errors: if the impossible occurs, then something fundamental is clearly wrong with the program. This is distinct from error handling: most error conditions are possible, although some may be extremely unlikely to occur in practice.

Using asserts for parameter checking is a frequently done error. You might rather change from assert->error in a few places instead, unless what you're checking can safly be **assert**ed to be impossible.

File: builtin/game/register.lua [hunk 2]
```diff
@@ -155,9 +192,28 @@ function core.register_item(name, itemdef)
...
+	core.registered_aliases[name] = nil
 	register_item_raw(itemdef)
+
+	if itemdef.can_place then
+		core.override_on_place(name, function(itemstack, placer, pt)
+			local pos = pt.under
+			local node = minetest.get_node(pos)
+			local def = minetest.registered_nodes[node.name]
+			if not def then
+				return
```
Comment (t4im) [inline builtin/game/register.lua]: Always return the itemstack in `on_place`. Or you make others crash when taking you up on the promise.

File: builtin/game/register.lua [hunk 3]
```diff
@@ -155,9 +192,28 @@ function core.register_item(name, itemdef)
...
+			end
+			if not def.buildable_to then
+				pos = pt.above
+				node = minetest.get_node(pos)
+			end
+			itemdef.can_place(vector.new(pos), node, itemstack, placer, pt)
```
Comment (t4im) [inline builtin/game/register.lua]: Always return the itemstack in `on_place`.

File: builtin/game/register.lua [hunk 4]
```diff
@@ -155,9 +192,28 @@ function core.register_item(name, itemdef)
...
-	core.registered_items[itemdef.name] = itemdef
-	core.registered_aliases[itemdef.name] = nil
+	core.registered_items[name] = itemdef
+	core.registered_aliases[name] = nil
 	register_item_raw(itemdef)
+
+	if itemdef.can_place then
+		core.override_on_place(name, function(itemstack, placer, pt)
+			local pos = pt.under
+			local node = minetest.get_node(pos)
```
Comment (t4im) [inline builtin/game/register.lua]: Doesn't handle CONTENT_IGNORE, might end up eating nodes.

File: builtin/game/register.lua [hunk 5]
```diff
@@ -155,9 +192,28 @@ function core.register_item(name, itemdef)
...
+			local def = minetest.registered_nodes[node.name]
+			if not def then
+				return
+			end
+			if not def.buildable_to then
+				pos = pt.above
+				node = minetest.get_node(pos)
```
Comment (t4im) [inline builtin/game/register.lua]: No CONTENT_IGNORE handling again.

File: builtin/game/register.lua [hunk 6]
```diff
@@ -155,9 +192,28 @@ function core.register_item(name, itemdef)
...
 
 	--core.log("Registering item: " .. itemdef.name)
-	core.registered_items[itemdef.name] = itemdef
-	core.registered_aliases[itemdef.name] = nil
+	core.registered_items[name] = itemdef
+	core.registered_aliases[name] = nil
 	register_item_raw(itemdef)
+
+	if itemdef.can_place then
+		core.override_on_place(name, function(itemstack, placer, pt)
```
Comment (t4im) [inline builtin/game/register.lua]: So you're overriding `on_place` when there is a `can_place`, but you don't call any preexisting `on_place`?

File: builtin/game/register.lua (hunk 4 above)
Comment (HybridDog) [inline builtin/game/register.lua]: but minetest.registered_nodes["ignore"] is nil, isn't it?

**Comment by t4im** [comment]
I just made a few unnecessary review comments, because i was a bit confused about the intention behind `override_on_place`, which actually doesn't allow one to override `on_place`? but instead is passed a `can_place`, and uses that to check whether or not to call the original `can_place`?

File: builtin/game/register.lua [hunk 7]
```diff
@@ -90,10 +90,34 @@ function core.register_entity(name, prototype)
...
 end
 
+
+function core.override_on_place(name, func)
+	local previous_on_place = core.registered_nodes[name].on_place
+	core.override_item(name, {
+		on_place = function(...)
+			if func(...) then
+				return previous_on_place(...)
+			end
```
Comment (t4im) [inline builtin/game/register.lua]: This would not return the itemstack if func returns false.

Comment (HybridDog) [inline builtin/game/register.lua]: fixed

File: builtin/game/register.lua (hunk 4 above)
Comment (t4im) [inline builtin/game/register.lua]: Oh, you're right. Nvm then. :-)

**Comment by HybridDog** [comment]
>override_on_place, which actually doesn't allow one to override on_place?

@est31 suggested calling it override_on_place, it actually changes it in a way to just disallow placement

File: doc/lua_api.txt [hunk 8]
```diff
@@ -2298,6 +2299,10 @@ These functions return the leftover itemstack.
     * Returns rating of the connect_to_raillike group corresponding to name
     * If name is not yet the name of a connect_to_raillike group, a new group id
     * is created, with that name
+* `minetest.override_on_place(name, func)`: overwrites the on_place of `name`
+    * if func doesn't return true, the previous on_place isn't executed
+* `minetest.override_on_punch(name, func)`: overwrites the on_punch of `name`
+    * if func doesn't return true, the previous on_punch isn't executed
```
Comment (est31) [inline doc/lua_api.txt]: Perhaps "original" is a better word than "previous"?

File: doc/lua_api.txt [hunk 9]
```diff
@@ -3109,6 +3114,10 @@ Definition tables
         --[[
         ^ Shall place item and return the leftover itemstack
         ^ default: minetest.item_place ]]
+        can_place = func(pos, node, itemstack, placer, pointed_thing),
+        --[[
+        ^ if it exists and returns true, the on_place^ is executed
```
Comment (est31) [inline doc/lua_api.txt]: one ^ too much

File: doc/lua_api.txt [hunk 10]
```diff
@@ -3246,6 +3255,10 @@ Definition tables
         on_punch = func(pos, node, puncher, pointed_thing), --[[
         ^ default: minetest.node_punch
         ^ By default: Calls minetest.register_on_punchnode callbacks ]]
+         can_punch = func(pos, node, puncher, pointed_thing),
+        --[[
+        ^ if it exists and returns true, the on_punch^ is executed
```
Comment (est31) [inline doc/lua_api.txt]: same here

File: builtin/game/register.lua [hunk 11]
```diff
@@ -338,25 +398,24 @@ function core.run_callbacks(callbacks, mode, ...)
 	if cb_len == 0 then
 		if mode == 2 or mode == 3 then
 			return true
-		elseif mode == 4 or mode == 5 then
+		end
```
Comment (est31) [inline builtin/game/register.lua]: why this change? This was perfectly fine the way it was.

**Comment by t4im** [comment]
May i suggest similar to the `allow_metadata_inventory_*` to use an "allow_" prefix to point out the intention behind the function? :)

File: doc/lua_api.txt (hunk 8 above)
Comment (HybridDog) [inline doc/lua_api.txt]: if minetest.override_on_punch is used multiple times, it doesn't take the original on_punch, but the one which became changed because minetest.override_on_punch was executed before

File: doc/lua_api.txt (hunk 9 above)
Comment (HybridDog) [inline doc/lua_api.txt]: it should be an arrow which is not an unicode one, which isn't readable by everyone, e.g. ↑

File: builtin/game/register.lua (hunk 11 above)
Comment (HybridDog) [inline builtin/game/register.lua]: if return is used, elseif does the same than another if, l guess another if works faster

**Comment by HybridDog** [comment]
allow_on_place sounds strange l think

**Comment by est31** [comment]
@t4im 's point is right, perhaps `override_can_place` then?

**Comment by t4im** [comment]
Better, yea :)

**Comment by HybridDog** [comment]
override can_place would mean that the node's can_place would be overwritten, it only overrides the on_place

**Comment by t4im** [comment]
Well technically it does not. It Wraps its call with a function making a decision on whether to precede with it. It's actually, when you think about it, a very bad behavior, which is reflected by the problem of naming it right.

You cannot override on_place with it at all, and you cannot really override can_place with it either. You can just wrap it in another (possibly quite deeply unnecessarily nested) call-deciding function.

**Comment by est31** [comment]
It isn't a full on_place replacement though, is it? And it is functionally equivalent than a can_place override, no?

**Comment by HybridDog** [comment]
>It's actually, when you think about it, a very bad behavior, which is reflected by the problem of naming it right.

l don't know what you mean by bad behaviour, it hardly happens that the function is used for one node often (>= 5 times), l guess.
> you cannot really override can_place with it either (but that seems closer to the intention)

>And it is functionally equivalent than a can_place override, no?

l actually just wanted to tell you that calling it "override_can_place" wouldn't fit because the PR adds a can_place (which offers pos and node) to nodes and that function changes the on_place (pos and node aren't given), not the can_place.
>It isn't a full on_place replacement though, is it?

the on_place is replaced with a different function, which executes the old on_place if the placing restriction function given returns … see https://github.com/minetest/minetest/pull/3226/files#diff-7b4242351dd3cf5b2a70c8aa9d12149bR94
ie it's wrapped

File: builtin/game/register.lua (hunk 11 above)
Comment (SmallJoker) [inline builtin/game/register.lua]: On return it 'jumps' to anothre line and continues with the code execution there, renaming `elseif` to `if` has no effect.

File: doc/lua_api.txt (hunk 8 above)
Comment (SmallJoker) [inline doc/lua_api.txt]: `isn't`
Using the long form `is not` like in `minetest.raillike_group(name)` makes it more standardized.

File: doc/lua_api.txt (hunk 9 above)
Comment (SmallJoker) [inline doc/lua_api.txt]: `Executes on_place when returning true`?

File: builtin/game/register.lua (hunk 11 above)
Comment (HybridDog) [inline builtin/game/register.lua]: but it makes it easier to understand that the function doesn't continue after the return

File: builtin/game/register.lua (hunk 4 above)
Comment (SmallJoker) [inline builtin/game/register.lua]: This does not look like a nil value to me:
https://github.com/minetest/minetest/blob/master/builtin/game/register.lua#L292

Comment (HybridDog) [inline builtin/game/register.lua]: SmallJoker is right

File: doc/lua_api.txt (hunk 9 above)
Comment (HybridDog) [inline doc/lua_api.txt]: yes

File: builtin/game/register.lua (hunk 11 above)
Comment (nerzhul) [inline builtin/game/register.lua]: @HybridDog let this else if alone.

Comment (HybridDog) [inline builtin/game/register.lua]: "else if" would need two "end"s

File: builtin/game/register.lua [hunk 12]
```diff
@@ -336,27 +406,22 @@ function core.run_callbacks(callbacks, mode, ...)
 	assert(type(callbacks) == "table")
 	local cb_len = #callbacks
 	if cb_len == 0 then
-		if mode == 2 or mode == 3 then
-			return true
-		elseif mode == 4 or mode == 5 then
-			return false
-		end
+		return (mode == 2 or mode == 3)
+				or (mode == 4 or mode == 5 or nil) and false
```
Comment (ShadowNinja) [inline builtin/game/register.lua]: This statement is almost indecipherable, please remove this change.

**Comment by HybridDog** [comment]
>Why? This is just what assert is for.

https://github.com/minetest/minetest/pull/3226#issuecomment-145370635

File: builtin/game/register.lua (hunk 12 above)
Comment (HybridDog) [inline builtin/game/register.lua]: Can l add a command describing it instead?

Comment (est31) [inline builtin/game/register.lua]: Code should be dead simple. I agree with @ShadowNinja.

Comment (HybridDog) [inline builtin/game/register.lua]: yes, returning nil there would be a mistake, it should continue

**Comment by HybridDog** [comment]
rebased

**Comment by ShadowNinja** [comment]
@HybridDog: There's a quote in that comment, but it has no refference.  It is true for C/C++ assertions, but Lua assertions are very different.  assert() is litteraly just `function assert(v, msg) if not v then error(msg or "assertion failed!") end return v end`.

**Comment by est31** [comment]
Please, state a reason why you close this.

**Comment by HybridDog** [comment]
l can't handle all my opened pull requests, e.g. l don't have enough time to change back the assert→error change
